### PR TITLE
Add create, last update timestamp to student entity #4476

### LINKED ISSUE #4476: Add create, last update timestamp to student entity
Last update time to be indexed.

### PR DESCRIPTION
**Pull request by LowWeiLin**
Fixes #4476

### CODE REVIEW

File: src/main/java/teammates/common/datatransfer/StudentAttributes.java [hunk 1]
```diff
@@ -67,6 +68,13 @@ public static UpdateStatus enumRepresentation(int numericRepresentation) {
     public String key = null;
 
     public UpdateStatus updateStatus = UpdateStatus.UNKNOWN;
+    
+    /*
+     * Creation and update time stamps. 
+     * Updated automatically in Student.java, jdoPreStore()
+     */
+    private Date created;
+    private Date lastUpdate;
```
Comment (damithc) [inline src/main/java/teammates/common/datatransfer/StudentAttributes.java]: `createdAt` `updatedAt`?

File: src/main/java/teammates/common/datatransfer/StudentAttributes.java [hunk 2]
```diff
@@ -357,4 +369,26 @@ public String getStudentStatus() {
...
+    }
+
+    public Date getLastUpdate() {
+        return lastUpdate;
+    }
+    
+    /**
+     * Should only be used for testing
+     **/
+    public void setCreated(Date created) {
```
Comment (damithc) [inline src/main/java/teammates/common/datatransfer/StudentAttributes.java]: May be we can add a `_NonProduction` suffix to indicate this method is not be used in production? The comment could be missed easily.

File: src/main/java/teammates/storage/api/StudentsDb.java [hunk 3]
```diff
@@ -364,15 +372,25 @@ public void updateStudentWithoutSearchability(String courseId, String email,
             String newName,
             String newTeamName, String newSectionName, String newEmail,
             String newGoogleID,
+            String newComments,
+            boolean keepUpdateTimestamp) throws InvalidParametersException,
+            EntityDoesNotExistException {
+        updateStudent(courseId, email, newName, newTeamName, newSectionName,
+                                        newEmail, newGoogleID, newComments, false, keepUpdateTimestamp);
+    }
```
Comment (damithc) [inline src/main/java/teammates/storage/api/StudentsDb.java]: The header comment needs to be updated to explain the meaning of last parameter.

File: src/main/java/teammates/storage/entity/Student.java [hunk 4]
```diff
@@ -93,6 +111,27 @@ public Student(String email, String name, String googleId, String comments, Stri
...
+        this.created = created;
+        setLastUpdate(created);
+    }
+    
+    public Date getLastUpdate() {
+        return lastUpdate;
+    }
+    
+    public void setLastUpdate(Date lastUpdate) {
+        if (keepUpdateTimestamp == false) {
```
Comment (damithc) [inline src/main/java/teammates/storage/entity/Student.java]: Use `if (!keepUpdateTimestamp)`? Or did you use this format to make it hard to miss?

File: src/test/java/teammates/test/driver/AssertHelper.java [hunk 5]
```diff
@@ -16,7 +16,19 @@
...
 
 public class AssertHelper {
-
+    
+    /**
+     * Assert date is now +- 1 min
+     */
+    public static void assertDateIsNow(Date date) {
+        assertDateWithinRange(date, new Date(new Date().getTime() - (1000 * 60)),
+                                    new Date(new Date().getTime() + (1000 * 60)));
```
Comment (damithc) [inline src/test/java/teammates/test/driver/AssertHelper.java]: See if there is a utility method (in a helper class) for getting dates/times with an offset.

**Comment by LowWeiLin** [comment]
Changes made

File: src/test/java/teammates/test/cases/storage/StudentsDbTest.java [hunk 6]
```diff
@@ -34,6 +35,53 @@ public static void setupClass() throws Exception {
...
+        
+        StudentAttributes student = studentsDb.getStudentForGoogleId(s.course, s.googleId);
+        assertNotNull(student);
+        
+        // Assert dates are now.
+        AssertHelper.assertDateIsNow(student.getCreatedAt());
+        AssertHelper.assertDateIsNow(student.getUpdatedAt());
+        
+        
+        ______TS("success : lastUpdated");
```
Comment (damithc) [inline src/test/java/teammates/test/cases/storage/StudentsDbTest.java]: `update lastUpdated`? The next test case says `keep lastUpdated`

**Comment by damithc** [comment]
What happens if we have to roll back to a previous version after deploying this version? The old code will work with entities created by this code?

**Comment by LowWeiLin** [comment]
I believe they will work, but some tests will fail due to the extra fields.

**Comment by damithc** [comment]
Do a manual test to make sure it does work (i.e. code can work when there are extra fields) :-)

**Comment by wkurniawan07** [comment]
Last time when I added the same field to comments, I added  null check for the new fields added. Should be working similarly here.

**Comment by LowWeiLin** [comment]
Performed a manual check, code can still work with the modified entities.

**Comment by damithc** [comment]
Yes @wkurniawan07 if the new field is used, we have to do a null check. I don't think it is used right now but @LowWeiLin may be we should add the null check to the getters?

**Comment by LowWeiLin** [comment]
The gettters return null if there is no timestamp. Should it return some other default value?
I was assuming that the null value will be handled when it is used.

**Comment by damithc** [comment]
Returning null is risky. I think we should return a default value if it is not set. e.g. 2011.01.01
It can be removed after we've done the full migration.

**Pull request merged**